### ISSUE TITLE
Allow creators to mute audio on their videos

### ISSUE THREAD

**Issue by o-fernandez** [GoogleForCreators/web-stories-wp#7730]
## Feature Description

Creators often upload videos with sound that they don't want in the final output of the story, allow them to easily set that the audio track of the video is muted during story playback. 

## Additional Context

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

**Comment by BrittanyIRL** [comment]
Feels related to this! https://github.com/google/web-stories-wp/issues/7371

**Comment by miina** [comment]
Per documentation, the video [`muted` setting is deprecated](https://amp.dev/documentation/components/amp-video/#muted-(deprecated)) and can't be configured directly. When `autoplay` is set, the video is automatically muted, otherwise not.

Currently, we're auto-playing all videos which means that all videos should be on mute by default.

@o-fernandez Thoughts?

**Comment by swissspidy** [comment]
@newmuis 👋  do you perhaps have a suggestion for us here from the format side regarding muted videos in stories, given that the `muted` attribute is considered deprecated on `amp-video`?

**Comment by newmuis** [comment]
I think we've had this feature request before.  I think it makes sense to tie to the existing `noaudio` attribute, but I'd leave it to @gmajoulet and @alanorozco

**Comment by alanorozco** [comment]
We've disabled `muted` for autoplay semantics, so in any case we should not use that attribute.

In my opinion, it's not advisable to encourage muting by markup. The audio track is sent over the wire, and wasted. It's much better to remove the audio track instead. There's many existing tools, even web apps, that can remove it.

**Comment by swissspidy** [comment]
Thanks for chiming in!

> In my opinion, it's not advisable to encourage muting by markup. The audio track is sent over the wire, and wasted. It's much better to remove the audio track instead. There's many existing tools, even web apps, that can remove it.

That's a good point. And given the current situation it seems to be the only viable solution. However, we can't expect creators to first run their videos through a separate tool to mute them. But I think we can offer this exact functionality directly in the editor (we're already using [ffmpeg.wasm](https://ffmpegwasm.github.io/)).

@o-fernandez WDYT about adding a button to the design panel that says "Remove audio from video" that will permanently remove the video's audio track.

The flow would be identical to the "Optimize video" button. Just instead of compressing the video it would just mute it. (or perhaps do both, to do them a favor)

**Comment by o-fernandez** [comment]
Agree with Pascal's assessment. We should allow for something within the editor to remove the audio track if the user wants to "mute" the video. 

We need to design the UX to ensure we add the option somewhere where it is easy for creators to find, and consider how to message it so that creators understand how the feature will work (e.g., the video will permanently lose the audio track).

**Comment by bmattb** [comment]
@tmagalhaes-huge Flagging that this requires UXD in order to be picked up in the next 2 sprints for v1.10

**Comment by swissspidy** [comment]
@o-fernandez Should we hide the pre-publish warning for missing captions & hide the captions panel itself for muted video? What if someone still wants to describe the video despite being muted?

**Comment by o-fernandez** [comment]
> @o-fernandez Should we hide the pre-publish warning for missing captions & hide the captions panel itself for muted video? What if someone still wants to describe the video despite being muted?

I think hiding the pre-publish checklist for captions makes sense, but we can keep the accessibility field for more descriptions and the ability to upload captions if the user wants (in case they want a silent video with captions). WDYT?

**Comment by spacedmonkey** [comment]
Having a think about this. I wonder this should be the course of action. 

On video elements, add option to "Convert to gif". This means it would convert the video into a video file without sound. Then add an option on gifs, to loop / not loop. Gif elements are videos without sound. Would if that makes more sense then adding option to mute a video.

**Comment by swissspidy** [comment]
If I recall our last conversation correctly, for the "convert the video into a video file without sound", we'd want to:

1. Add a button to the design panel for muting the video.
2. Add a new method to `useProcessVideo` and `useFFmpeg` to remove sound from video and upload that video to WordPress
3. Swap out the video on the canvas with the new one without sound
  The new video will have necessary `isMuted` properties everywhere. That means updating the `Resource` type.
4. Keep the original video unchanged & visible in the media library  
  This is different from video optimization (where we hide the original), because there could be a case where you want to use the video _with_ sound in some other place.

I think this option should be called "Mute video" or similar, not "convert to GIF". For a couple of reasons:

* Users know GIFs as looped, animated images, not still videos that can be looped or not looped. GIFs always loop.  
  Offering an option to convert a video to a GIF when all you want is to mute your video would be very confusing.
* The UI for GIFs in the editor is different from videos. They don't have a loop option and there is no way to set/change the poster image manually.  
  In other words, the GIF element _appears_ like an image to the user, not a video. That's our goal.

Regardless of that, let's focus on the 4 steps above first as the MVP for this functionality. Whether this says GIF or video in the end can be decided afterwards. First we need a way to remove the sound.

**Comment by spacedmonkey** [comment]
Thanks for the feedback, I was just thinking about load / validating my thinking. I agree, steps 1-4 ( which we agree in sprint planning ), should be the course forward. 

> Users know GIFs as looped, animated images, not still videos that can be looped or not looped. 

Yeah, that is what the user thinks, making it confusing, I guess. Now we convert gifs under the hood to mp4, there isn't a difference, but to the user it is different.

**Comment by spacedmonkey** [comment]
@swissspidy Should we add some check in on upload, to mark an uploaded video as not having sound? What if a user uploads a video without sound? Having to click a button to say, "remove sound from video" when the user knows that video already doesn't have sound, is a pain. 

Also thinking about having a piece of post meta, to store, the "isMuted" field in WP.

**Comment by spacedmonkey** [comment]
I put together a POC of checking to see if video has video in PHP level. 

https://github.com/google/web-stories-wp/commit/1b6fc422104ab0bb04f81dd3f1a3e911d0d07417

**Comment by swissspidy** [comment]
> @swissspidy Should we add some check in on upload, to mark an uploaded video as not having sound? What if a user uploads a video without sound? Having to click a button to say, "remove sound from video" when the user knows that video already doesn't have sound, is a pain.
> 
> Also thinking about having a piece of post meta, to store, the "isMuted" field in WP.

Makes sense to store this information also on the server 👍 

> I put together a POC of checking to see if video has video in PHP level.

The discussion on #7371 comes to mind here.

There, my worry about using `wp_get_attachment_metadata` was that the information might not be reliable. (If metadata extraction fails on some servers, then there's of course no information about audio channels, so we might wrongly assume it's muted)

But then again, all we do with this information is hiding some warnings in the UI, so I think it's acceptable to use it.

So let's proceed with that 👍

**Comment by spacedmonkey** [comment]
In 1b6fc42 I only get audio track data for mp4/webm files. In my testing, ALL of files I tested has correct information for these file types. Again, this only for older video files that have been uploaded. I think we should set the flag to false on file upload. 

I was wondering, if I should do a check to see if the file has audio in Javascript and set the post meta to true/ false after upload, a little similar to poster generation. 

I proptypes a new util to find out if a video has audio or not. 

https://jsfiddle.net/ws2jzrvL/1/

**Comment by csossi** [comment]
Verified in QA
